### PR TITLE
feat: use type origin map to get correct types when upgrading package

### PR DESCRIPTION
**Pull request by karlwuest**
## Description

Allow specifying a Walrus package ID (e.g. for package upgrades) and use the package to get the correct fully qualified types for all move structs defined in the Walrus package.

This currently still assumes that all packages (`wal`, `wal_exchange`, and `walrus`) are published as a single package. I will separate this in a follow up (see WAL-425).

Contributes to WAL-346

## Test plan

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.
For each box you select, include information after the relevant heading that describes the impact of your changes that
a user might notice and any actions they must take to implement updates. (Add release notes after the colon for each item)

- [ ] Storage node:
- [ ] Aggregator:
- [ ] Publisher:
- [ ] CLI:

### CODE REVIEW

File: crates/walrus-service/client_config_example.yaml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 system_object: 0xa2637d13d171b278eadfa8a3fbe8379b5e471e1f3739092e5243da17fc8090eb
 staking_object: 0xca7cf321e47a1fc9bfd032abc31b253f5063521fd5b4c431f2cdd3fee1b4ec00
+walrus_package: 0xa9b00f69d3b033e7b64acff2672b54fbb7c31361954251e235395dea8bd6dcac
 exchange_object:
-- 0xa9b00f69d3b033e7b64acff2672b54fbb7c31361954251e235395dea8bd6dcac
 - 0x26a8a417b553b18d13027c23e8016c3466b81e7083225436b55143c127f3c0cb
+- 0xaf3819c82de2e0257c0cc2177dfce6432efa42ca6c04c0b774dbb3c5ca2573cd
```
Comment (giac-mysten) [inline crates/walrus-service/client_config_example.yaml:6]: question: is this change necessary?

File: crates/walrus-sui/src/client/read_client.rs [hunk 2]
```diff
@@ -376,6 +398,86 @@ impl SuiReadClient {
             })
     }
 
+    /// Get the [`StorageNodeCap`] object associated with the address.
```
Comment (giac-mysten) [inline crates/walrus-sui/src/client/read_client.rs:401]: Question: any specfic reason to move these functions  to methods on the read client? just cleanup?

File: crates/walrus-sui/src/client/read_client.rs [hunk 3]
```diff
@@ -706,6 +808,30 @@ pub async fn get_system_package_id(
     Ok(object_pkg_id)
 }
 
+/// Gets the type origin map for a given package.
+async fn type_origin_map_for_package(
```
Comment (giac-mysten) [inline crates/walrus-sui/src/client/read_client.rs:812]: nice!

File: crates/walrus-sui/src/contracts.rs [hunk 4]
```diff
@@ -115,8 +118,10 @@ pub struct StructTag<'a> {
 }
 
 impl<'a> StructTag<'a> {
-    /// Returns a Move StructTag for the identified struct, within the published contract module.
-    pub fn to_move_struct_tag(
+    /// Returns a [MoveStructTag] for the identified struct with the given package ID.
```
Comment (giac-mysten) [inline crates/walrus-sui/src/contracts.rs]:
```suggestion
    /// Returns a [`MoveStructTag`] for the identified struct with the given package ID.
```

File: crates/walrus-sui/src/contracts.rs [hunk 5]
```diff
@@ -132,6 +137,19 @@ impl<'a> StructTag<'a> {
             type_params: type_params.into(),
         })
     }
+
+    /// Converts a [StructTag] to a [MoveStructTag] using the matching package ID from the given
```
Comment (giac-mysten) [inline crates/walrus-sui/src/contracts.rs]:
```suggestion
    /// Converts a [`StructTag`] to a [`MoveStructTag`] using the matching package ID from the given
```

File: crates/walrus-service/client_config_example.yaml (hunk 1 above)
Comment (karlwuest) [reply crates/walrus-service/client_config_example.yaml:6]: That's automatically generated and I think it changed because of the order in which the random ObjectIDs are generated.

File: crates/walrus-sui/src/client/read_client.rs (hunk 2 above)
Comment (karlwuest) [reply crates/walrus-sui/src/client/read_client.rs:401]: Yes, mainly cleanup. I think, we used to use (some of) them separately during the setup, before we actually constructed a read client. But I think this is the right place for them (it also simplifies their signatures a bit).

File: crates/walrus-sui/src/client/read_client.rs (hunk 3 above)
Comment (karlwuest) [reply crates/walrus-sui/src/client/read_client.rs:812]: thanks!
